Ant Mission 1

Cheesy0314 opened the pull request:
Created new PR with just ant mission 1, last request had multiple missions in it.

Pulled map.bin and map.yaml from #15213, map.png from cnc-wiki and rescaled.

This is an attempt to complete part of #4989

Commit:
Add ant mission 1

Smittytron reviewed (commented):
This looks promising! Here are some initial thoughts after playing it through.

- Was the timer really 30 full minutes? That's a long time, I bet the clock ticked faster.

- Because the base buildings are allied to the player, their health bars will show to anyone who has bars set to show on damage or always show.

- The player starts with 1500 cash in the original. Also, destroying the oil derrick yielded a money crate.

- Snipers need to be disabled.

- The ant colors will need tweaking. The scouts were black, fire ants were orange, and warriors were red.

- Do we want the openra flamethrower or the fireballs for the fire ants? I'm leaning fireballs.

- Change the commit title to something like 'Add Ant Mission 1'

Cheesy0314 commented:
Ok, so far:
- Disabled Snipers
- Set starting funds to 1500
- Changed building owner allies so health bar wouldn't display
- Changed commit title

TODO:
- Attempt to tweak colors of ants
- Change fireball animation
- Possibly decrease time (Although original **was** 30 mins)

johnweven commented on mods/ra/maps/ant-01/ant-01.lua:
```diff
@@ -0,0 +1,310 @@
+FireAnts = {"fireant"}
+WarriorAnts = {"ant"}
+AlliedForces = {"2tnk","2tnk","mcv"}
+ChooperTeam = {"e1r1","e1r1","e2","e2","e1r1"}
```
Was it supposed to say Chopper? Or was it intended to say Chooper?

Cheesy0314 commented:
No haha meant chopper, hit the "o" twice instead of "p", good catch!


>I am very glad to hear that someone was willing to do the scripts! 
@johnweven  Yes I saw your branch and figured I'd take a crack at it! You did the part I didn't want to do so it was easy enough after that. Plus, had a few slow days at work so it made the day seem quicker.

So update: 
Fixe Chooper spell mistake
Changed fire animation, needs more work
Couldn't fix color, not sure why player color isn't changing
Changed Timer start based on difficulty

TODO:
Finalize last changes

Smittytron commented on mods/ra/maps/ant-01/rules.yaml:
```diff
@@ -0,0 +1,95 @@
+	Mobile:
+		Speed: 80
+	Armament:
+		Weapon: Flamer
```
I'd just use `FireballLauncher` as the weapon.

Smittytron commented on mods/ra/maps/ant-01/rules.yaml:
```diff
@@ -0,0 +1,95 @@
+	Armament:
+		Weapon: Flamer
+	Health:
+		HP: 70000
```
Style nit:  Use empty lines between each rule

Smittytron commented on mods/ra/maps/ant-01/rules.yaml:
```diff
@@ -0,0 +1,95 @@
+		HP: 80000
+MEDI:
+	Valued:
+		Cost: 500
```
I'd take this out. The default medic price shouldn't hurt anything.

Smittytron commented on mods/ra/maps/ant-01/rules.yaml:
```diff
@@ -0,0 +1,95 @@
+		Prerequisites: ~disabled
+MNLY:
+	Buildable:
+		Prerequisites: ~techlevel.low
```
This override shouldn't be necessary.

Smittytron commented on mods/ra/maps/ant-01/rules.yaml:
```diff
@@ -0,0 +1,95 @@
+		Prerequisites: ~disabled
+SNIPER:
+	Buildable:
+		Prerequisites: ~disabled
```
Snipers are disabled by default. Removing the custom override should have been enough to take it out.

Smittytron commented on mods/ra/maps/ant-01/rules.yaml:
```diff
@@ -0,0 +1,95 @@
+		Scripts: ant-01.lua
+	MissionData:
+		BackgroundVideo: antintro.vqa
+		Briefing: We've lost contact with one of our outposts. Before it went off-line, we recieved a brief communique about giant ants. We're unsure what to make of this report, so we want you to investigate.  \n\nScout the area, bring the outpost back on-line, and report your findings. If there is a threat, reinforcements will be sent in to help you.  \n\nKeep the base functional and radio contact op ...
```
Move player and world to the top. This is how the other missions are sorted.

Cheesy0314 commented:
Made requested changes, removed unused weapons.yaml

Fixed bug in lua scrip

MustaphaTR commented on mods/ra/maps/ant-01/ant-01.lua:
```diff
@@ -0,0 +1,310 @@
+TimerTicks = DateTime.Minutes(30)
+ticks = TimerTicks
+
+WorldLoaded = function()
```
World Loaded should go to end and init objectives above it.

MustaphaTR commented on mods/ra/maps/ant-01/ant-01.lua:
```diff
@@ -0,0 +1,310 @@
+        AntSendFunc(direction, amount, "ant")
+end
+
+
```
Double new line.

Smittytron commented:
I was able to override the colors by removing the campaign-palettes, but I don't know if that's the right way to go about it. Would @abcdefg30 or @MustaphaTR know if that's an acceptable workaround?

MustaphaTR commented:
What is wrong with the colors again? I think they are correct.

Also something i noticed. You can select and make the chinook stay. Preferably it should be unselectable so you can't mess with it.

Cheesy0314 commented on mods/ra/maps/ant-01/ant-01.lua:
```diff
@@ -0,0 +1,310 @@
+InsertionHelicopterType = "tran"
+InsertionPath = { waypoint12.Location, waypoint0.Location }
+baseDiscovered = false
+AlliedBase = {Actor99, Actor100, Actor101, Actor102, Actor103, Actor104, Actor105, Actor106, Actor107, Actor129}
```
Yeah I meant to name them and never did, good call

Cheesy0314 commented:
Ok, update with better coding practices!

Cheesy0314 commented:
Ok, removed generic names/visibility

Modified ant weapons to make fireball a little less destructive and warrior ant mandible a little more destructive

Cheesy0314 commented:
Yeah I think that’s a fair enough idea, I was still trying to decide the best way to scale the waves with that time frame.

Smittytron commented:
Personally I'd go with harder waves instead of more time (if you tackle difficulty). A mission where you just sit there and defend can only be interesting for so long.

Smittytron commented:
Yeah, reaperrr mathed it out to 18 mins so I'd go with that.

Cheesy0314 commented:
Scaled the timer appropriately, also changed ants per wave based on difficulty.

Easy == regular amount
Normal == doubled amount
Hard == tripled amount

Smittytron commented:
Sorry to keep asking for changes; I think we're getting close. Always a lot of work with these missions. The good news is most of this can be applied to the other three.

> What is wrong with the colors again?

@MustaphaTR  The red warrior ants look right, but the charcoal/black scout ants are currently lime green and the orange fire ants are dark red.

Cheesy0314 commented:
> Sorry to keep asking for changes;

@Smittytron It's cool, I'm well acquainted with the code review process. As for the colors I'll try again to mess around with them.

pchote commented:
A couple of comments without reading the discussion here in depth:

1. We should pull these ant definitions out to a common file in the rules directory.  There's no point duplicating them across multiple maps.

2. > The ant colors will need tweaking. The scouts were black, fire ants were orange, and warriors were red.
   
   The original game did color ranges by reindexing the player indices to a different range in the palette.     You can use our `IndexedPlayerPalette` to do the same thing and obtain accurate results without  having to guess at RGB values.

Cheesy0314 commented:
Ok, I can move the ant definitions which is what I originally planned on doing but didn't want to change anything in rules/ on my own.

I'll read up on how to use IndexedPlayerPalette for the ant colors too, thank you @pchote

Cheesy0314 commented:
Alright, I **_think_** I have the colors sorted out with using `IndexedPlayerPalette`

pchote commented:
In theory you should be able to keep a single `Ant` player by defining different player palettes for each unit.  Something like:
```yaml
^Palettes:
	IndexedPlayerPalette@scouts:
		BaseName: scoutant
		BasePalette: player
		RemapIndex: 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95
		PlayerIndex:
			Ants: 133, 134, 135, 136, 137, 138, 139, 140, 141, 141, 142, 142, 143, 143, 143, 143
		AllowModifiers: True
	IndexedPlayerPalette@warriors:
		BaseName: warriorant
		BasePalette: player
		RemapIndex: 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95
		PlayerIndex:
			Ants: 229, 230, 231, 232, 233, 234, 235, 8, 236, 237, 238, 239, 221, 222, 223, 223
		AllowModifiers: True
	IndexedPlayerPalette@fireguy:
		BaseName: fireant
		BasePalette: player
		RemapIndex: 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95
		PlayerIndex:
			Ants: 158, 158, 159, 159, 229, 229, 216, 216, 217, 217, 218, 218, 221, 222, 223, 223
		AllowModifiers: True
```

and then on each ant type:
```yaml
RenderSprites:
	PlayerPalette: fireant # etc
```

Cheesy0314 commented:
@pchote Oh yeah that's way easier thank you

Cheesy0314 commented:
Forgot to write that I update this, reduced ant teams down to one player and used IndexedPlayerColors to set each ant type appropriately.

Smittytron commented:
Gave it another play-through:

- Cash ticks are now grey instead of the player color. 
- It doesn't seem to throw anything but scout ants at the player.
- Couldn't for the life of me beat it on hard. I'd look into scaling difficulty with frequency of waves or waves with harder ant compositions. You can also use triggers to upgrade the ant composition to something harder later in the mission.

Cheesy0314 commented:
Ok, altered ant waves to scale frequency only, hard sends warrior ants or fire ants not scout.

I fixed player color for credit ticks, and I fixed the issue with only scout ants being sent. (I accidentally removed the code that picks other ant types, whoops!)

Smittytron commented:
Difficulty is close. I was able to last for 10 mins on hard this time. I experimented adding a $2000 crate to see if it would give me the scaling needed, and I was able to win this way.

The money crate was in the original and was revealed by destroying the oil pump below the base. If you want to add it, an example is in #14972. You can have the crate spawn to the side, or add an override rule that removes the oil pump on destruction.

Cheesy0314 commented:
@Smittytron Ok I added the money crate to spawn next to the pump

Smittytron reviewed (dismissed):
Difficulty was the last issue I saw. LGTM now.

reaperrr commented on mods/ra/maps/ant-01/rules.yaml:
```diff
@@ -0,0 +1,128 @@
+		BasePalette: player
+		RemapIndex: 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95
+		PlayerIndex:
+			AntMan: 133, 134, 135, 136, 137, 138, 139, 140, 141, 141, 142, 142, 143, 143, 143, 143
```
Those are the wrong colors, the original used
> 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 123, 123, 123, 123

Well, not sure about the exact distribution, I don't have the original palette.cps here, but definitely those 12 colors.

reaperrr commented on mods/ra/rules/infantry.yaml:
```diff
@@ -725,3 +725,46 @@ Ant:
+	Armament:
+		Weapon: AntFireball
+	Health:
+		HP: 70000
```
To match the original, this should be 7500, but with
```
	Armor:
		Type: Heavy
```

That will make at least bunkers and grenades more effective (and rockets/cannons in later missions).

reaperrr commented on mods/ra/rules/infantry.yaml:
```diff
@@ -725,3 +725,46 @@ Ant:
+		Name: Scout Ant
+		GenericVisibility: none
+	Health:
+		HP: 60000
```
This should be 8500, but with
```
	Armor:
		Type: Light
```

reaperrr commented on mods/ra/rules/infantry.yaml:
```diff
@@ -725,3 +725,46 @@ Ant:
+	Mobile:
+		Speed: 50
+	Health:
+		HP: 80000
```
And this should be 12500, with
```
	Armor:
		Type: Heavy
```

reaperrr commented on mods/ra/weapons/other.yaml:
```diff
@@ -170,6 +170,36 @@ Mandible:
 
+MandibleHeavy:
+	Inherits: Claw
+	ReloadDelay: 15
```
The ant attack sound is missing. Not sure how annoying it might get due to the low reload delay, though.

Sound name is `antbite`.

Same goes for `mandible` and `AntiFireball`.

Cheesy0314 replied:
Ok, I tested and it's definitely annoying. How should I handle this?

abcdefg30 replied:
What did the original do? Was it that annoying there as well?

reaperrr replied:
Not sure, I think in the original the sound was tied to the attack animation, which was slower than the hypothetical ROF.

We could try giving the mandible weapons Burst: 2, with BurstDelays set to 14 (needs to be lower than ReloadDelay) so players won't notice, and then use `StartBurstReport` instead of Report, so it only plays every 2nd attack.

Cheesy0314 replied:
@reaperrr Ok, I tried it out and it worked out nicely

Smittytron commented:
Tested with the new ants, and hard is back to 'almost beatable'. For scaling, I tried giving the crate 2000 instead of 500 and was able to survive much longer, but when all you face is ants with heavy armor you're going to bleed out. (grenadiers won't hit them until they stop moving to attack, and they kill one infantry per attack)

You can try giving the player more starting money for better scaling, but I think we might have to go with dialed back attacks, or maybe mix a few grey ants back in.

Cheesy0314 commented:
OK I reintroduced scout ants to hard and upped the amount of money in the crate. I made more ants possible earlier but delayed fire ants an extra minute

Smittytron commented:
Looks like you took warrior ants completely out. I'd make and upgrade trigger and have scout ants become warriors 4 minutes in on hard. 10 or so for Normal. I think that'll be enough time for the player to get established.

Cheesy0314 commented:
OK set the ant time logic for warrior ants, also delayed fire ants a minute or so

pchote commented:
What is the review status here @abcdefg30, @Smittytron, @reaperrr?

abcdefg30 commented:
There are a bunch of style issues here. @Cheesy0314 would you mind if I just fixed them and added my fixes to your commit (this will probably take less time than explaining each of them)? I can also file a PR with my suggestions towards your fork if you want that.

pchote commented:
Any progress here @abcdefg30?

matjaeck commented:
Just stumbled across https://www.pcgamer.com/red-alerts-ant-missions/ fyi

reaperrr commented:
Sorry for the long silence on this.

I've taken the liberty to push some fixes and difficulty adjustments to the branch, to save some time and hopefully get this merged quickly now.

Fixes:
- Fixed that the switch from occasional scouts to only warriors would only work on `normal`, and not `hard` (the `if` and `elsif` previously both checked for `DifficultySetting == "normal"`).
- Removed a few redundant empty lines from the lua scripts.

Difficulty adjustments:
1) Reduced max ants from 
  - 5 to 3 on easy, and
  - 6 to 5 on normal.
Note: This doesn't reduce difficulty as much as it seems, because a) fire ants remain untouched and b) due to the random factors, it only reduces the worst-case size of ant attacks towards the end.
2) Switch from random scouts to "only fire- and warrior ants" now happens 4 minutes later on `normal` (after 12 instead of after 8 mins), but 2 minutes earlier on `hard` (after 6 instead of 8 minutes).

👍 from me, @abcdefg30 if you still plan to adress those style issues it would be great if you could do that soon (or comment if you'd rather file a follow-up later).

Smittytron commented:
The spawn points for the ants should be at the edges of the map.
